gdal/frmts/jpegxr/jpegxrdataset.cpp

```diff
@@ -0,0 +1,1202 @@
+        if (config.GetQuality() == QUALITY_LOSSLESS)
+        {
+            wmiSCP.uiDefaultQPIndex = 1;
+        }
+        else
+        {
+            PKPixelInfo pi;
+            memset(&pi, 0, sizeof(PKPixelInfo));
+            pi.pGUIDPixFmt = &pImageEncode->guidPixFormat;
+            ERR nErr = PixelFormatLookup(&pi, LOOKUP_FORWARD);
+            if (nErr != WMP_errSuccess)
+                throw JPEGXRException(nErr);
+
+            wmiSCP.olOverlap = config.GetOVERLAP();
+
+            if (IsEqualGUID(pImageEncode->guidPixFormat, GUID_PKPixelFormatBlackWhite))
+            {
+                // TODO: B&W to be tested
+                //CPLAssert(pi.bdBitDepth == BD_1);
+                //wmiSCP->uiDefaultQPIndex = (U8)(8 - 5.0f * iq_float + 0.5f);
+                throw JPEGXRException(WMP_errUnsupportedFormat);
+            }
+            else
+            {
+                // Microsoft in JxrEncApp.c noted:
+                // Remap [0.8, 0.866, 0.933, 1.0] to [0.8, 0.9, 1.0, 1.1]
+                // to use 8-bit QP table (0.933 == Photoshop JPEG 100)
+                float fQuality = config.fQuality;
+                if (fQuality > 0.8f &&
+                    pi.bdBitDepth == BD_8 &&
+                    wmiSCP.cfColorFormat != YUV_420 &&
+                    wmiSCP.cfColorFormat != YUV_422)
+                {
+                    fQuality = 0.8f + (fQuality - 0.8f) * 1.5f;
+                }
+
+                int const qi = static_cast<int>(10.f * fQuality);
+                float const qf = 10.f * fQuality - static_cast<float>(qi);
+                int* pQP = nullptr;
+                if (wmiSCP.cfColorFormat == YUV_420 || wmiSCP.cfColorFormat == YUV_422)
+                {
+                    pQP = QP::BD8_420[qi];
+                }
+                else
+                {
+                    if (pi.bdBitDepth == BD_8)
+                        pQP = QP::BD8[qi];
+                    else if (pi.bdBitDepth == BD_16)
+                        pQP = QP::BD16[qi];
+                    else if (pi.bdBitDepth == BD_16F)
+                        pQP = QP::BD16f[qi];
+                    else
+                        pQP = QP::BD32f[qi];
+                }
+                CPLAssert(pQP);
+
+                wmiSCP.uiDefaultQPIndex    = (U8)(0.5f + pQP[0] * (1.f - qf) + (pQP + 6)[0] * qf);
+                wmiSCP.uiDefaultQPIndexU   = (U8)(0.5f + pQP[1] * (1.f - qf) + (pQP + 6)[1] * qf);
+                wmiSCP.uiDefaultQPIndexV   = (U8)(0.5f + pQP[2] * (1.f - qf) + (pQP + 6)[2] * qf);
+                wmiSCP.uiDefaultQPIndexYHP = (U8)(0.5f + pQP[3] * (1.f - qf) + (pQP + 6)[3] * qf);
+                wmiSCP.uiDefaultQPIndexUHP = (U8)(0.5f + pQP[4] * (1.f - qf) + (pQP + 6)[4] * qf);
+                wmiSCP.uiDefaultQPIndexVHP = (U8)(0.5f + pQP[5] * (1.f - qf) + (pQP + 6)[5] * qf);
+            }
+        }
+
+        // TODO: Tiling
+    }
+
+    void Write(GByte* pabyData, int nXStride, int nYSize)
```

Comment:
@rouault perhaps it's just unfortunate style of `PKImageEncode_WritePixels`, `PKImageEncode_EncodeContent_Init`, `PKImageEncode_EncodeContent_Encode` and other functions of jxrlib API, or they indeed expect modifiable chunk.
eg.

```
 PKImageEncode_WritePixels(
    PKImageEncode* pIE,
    U32 cLine,
    U8* pbPixels,
    U32 cbStride)
```